src/datasources/cache/cache.first.data.source.spec.ts

```diff
@@ -0,0 +1,134 @@
+    const data = JSON.parse(faker.datatype.json());
+    mockNetworkService.get.mockImplementation((url) => {
+      switch (url) {
+        case targetUrl:
+          return Promise.resolve({ data });
+        default:
+          return Promise.reject(`No matching rule for url: ${url}`);
+      }
+    });
+
+    const actual = await cacheFirstDataSource.get(cacheDir, targetUrl);
+
+    expect(actual).toEqual(data);
+    expect(mockNetworkService.get).toHaveBeenCalledTimes(1);
+  });
+
+  it('should return the cached data without calling the underlying network interface', async () => {
+    const cacheDir = new CacheDir(faker.random.word(), faker.random.word());
+    const rawJson = faker.datatype.json();
+    fakeCacheService.set(cacheDir, rawJson);
+    mockNetworkService.get.mockImplementation((url) =>
+      Promise.reject(`Unexpected request to ${url}`),
+    );
+
+    const actual = await cacheFirstDataSource.get(
+      cacheDir,
+      faker.internet.url(),
+    );
+
+    expect(actual).toEqual(JSON.parse(rawJson));
+    expect(mockNetworkService.get).toHaveBeenCalledTimes(0);
+  });
+
+  it('should cache 404 errors coming from the network', async () => {
+    const targetUrl = faker.internet.url();
+    const cacheDir = new CacheDir(faker.random.word(), faker.random.word());
+    const expectedError = new NetworkResponseError(404);
+    mockNetworkService.get.mockImplementation((url) => {
+      switch (url) {
+        case targetUrl:
+          return Promise.reject(expectedError);
+        default:
+          return Promise.reject(`No matching rule for url: ${url}`);
+      }
+    });
+
+    await expect(
+      cacheFirstDataSource.get(cacheDir, targetUrl),
+    ).rejects.toThrowError(expectedError);
+
+    expect(mockNetworkService.get).toHaveBeenCalledTimes(1);
+    expect(fakeCacheService.keyCount()).toBe(1);
+  });
+
+  it('should return a 404 cached error without a second call to the underlying network interface', async () => {
+    const targetUrl = faker.internet.url();
+    const cacheDir = new CacheDir(faker.random.word(), faker.random.word());
+    const expectedError = new NetworkResponseError(404);
+    mockNetworkService.get.mockImplementation((url) => {
+      switch (url) {
+        case targetUrl:
+          return Promise.reject(expectedError);
+        default:
+          return Promise.reject(`No matching rule for url: ${url}`);
+      }
+    });
+
+    await expect(
+      cacheFirstDataSource.get(cacheDir, targetUrl),
+    ).rejects.toThrowError(expectedError);
+
+    await expect(
+      cacheFirstDataSource.get(cacheDir, targetUrl),
+    ).rejects.toThrowError(expectedError);
+
+    expect(mockNetworkService.get).toHaveBeenCalledTimes(1);
+    expect(fakeCacheService.keyCount()).toBe(1);
+  });
+
+  it('should cache errors with the expected TTL', async () => {
+    const mockCache = jest.mocked({
+      get: jest.fn(),
+      set: jest.fn(),
+      delete: jest.fn(),
+    } as unknown as ICacheService);
```

Comment:
Suggestion: maybe we can have a `getTTL` on the `FakeCacheService` so you wouldn't need to mock another object just for this test.

Let me know what you think!

Reply:
Do you mean having something like getting the TTL given a key (similar to redis [TTL function](https://redis.io/commands/ttl/))? That would be nice, I had been thinking about something like that also, but I suspect we would need to change the data representation in `FakeCacheService` to save the TTL somewhere, right? So in that case `Record<string, Record<string, any>>` doesn't fit anymore and we would need something like `Record<Record<string, number>, Record<string, any>>`, or use tuples, to store the TTL there. And this would affect the current `FakeCacheService` functions as well, so I decided to simply use the mock approach in the test class. Do you think is it better to go through the other path?

Reply:
> 

Ok since we have a single test that checks the TTL at the moment I would say that this change is not needed right now and we can tackle it later if we write more tests around the TTL part 👍